engine/src/srvcgi.cpp

```diff
@@ -595,11 +595,125 @@ static void cgi_fetch_valueref_for_key(MCVariable *p_variable, MCNameRef p_key,
+            || !MCNameCreate(*t_key_str, &t_key)
+            || !t_path . Push(MCValueRetain(*t_key)))
+        return false;
+
+    t_fetched_value = p_variable -> getvalueref(t_path . Ptr(), t_path . Size(), false);
+
+    uindex_t t_subkey_start, t_subkey_end;
+
+    // Initialise t_subkey_end to find the first subkey after the initial key
+    t_subkey_end = t_key_end;
+
+    while (MCStringFirstIndexOfChar(t_raw_key_str, '[', t_subkey_end, kMCStringOptionCompareExact, t_subkey_start))
+    {
+        // Fetch the value at the current path.
+        t_fetched_value = p_variable -> getvalueref(t_path . Ptr(), t_path . Size(), false);
+
+        // The subkey starts after the '['
+        t_subkey_start++;
+
+        if (!MCStringFirstIndexOfChar(t_raw_key_str, ']', t_subkey_start, kMCStringOptionCompareExact, t_subkey_end))
+            return false;
+
+        MCAutoStringRef t_subkey_str;
+        MCNewAutoNameRef t_subkey;
+        // We create the subkeys as needed (unnamed keys are translated to
+        //  indices).
+        if (t_subkey_end == t_subkey_start)
+        {
+            // Subkey with no name: we must use an index
+            uindex_t t_index;
+
+            // getvalueref returns kMCEmptyString if the key path does not lead to
+            //  an existing value.
+            if (t_fetched_value == kMCEmptyString
+                    || !MCValueIsArray(t_fetched_value))
+            {
+                // Not an array: we put the value at the index 1
+                t_index = 1;
+            }
+            else if (MCArrayIsSequence((MCArrayRef)t_fetched_value))
+            {
+                // Sequence: we put after the last element
+                t_index = MCArrayGetCount((MCArrayRef)t_fetched_value) + 1;
+            }
+            else
+            {
+                // We get the first index available
+                for (t_index = 1; p_variable -> getvalueref(t_path.Ptr(), t_path.Size(), false) != kMCEmptyString; ++t_index)
+                    ;
+            }
+
+            if (!MCStringFormat(&t_subkey_str, "%u", t_index))
+                return false;
+        }
+        else
+        {
+            // We have a named key, we just add the name to the path.
+            if (!MCStringCopySubstring(t_raw_key_str, MCRangeMake(t_subkey_start, t_subkey_end - t_subkey_start), &t_subkey_str))
+                return false;
+        }
+
+        // Create a name from the subkey string, and append it to the path
+        if (!MCNameCreate(*t_subkey_str, &t_subkey)
+                || !t_path . Push(MCValueRetain(*t_subkey)))
+            return false;
+    }
+
+    // Store the value at the built key path - setvalueref will take care of
+    //  creating subarrays if needed.
+    bool t_success;
+    t_success = p_variable -> setvalueref(t_path . Ptr(), t_path . Size(), false, p_value);
+
+    // Clean keys array
+    t_path . Delete();
```

Comment:
I thought an `MCAutoArray()` automatically self-deletes? I don't think this (or the `t_success` variable) is necessary...?